hal/armv8m/string.c

```diff
@@ -0,0 +1,256 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * HAL basic routines
+ *
+ * Copyright 2017, 2022 Phoenix Systems
+ * Copyright 2001, 2005-2006 Pawel Pisarczyk
+ * Author: Pawel Pisarczyk, Artur Wodejko, Aleksander Kaminski, Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "../string.h"
+
+
+void hal_memcpy(void *dst, const void *src, unsigned int l)
+{
+	/* clang-format off */
+	__asm__ volatile
+	(" \
+		orr r3, %0, %1; \
+		lsls r3, r3, #30; \
+		bne 2f; \
+	1: \
+		cmp %2, #4; \
+		ittt hs; \
+		ldrhs r3, [%1], #4; \
+		strhs r3, [%0], #4; \
+		subshs %2, #4; \
+		bhs 1b; \
+	2: \
+		cmp %2, #0; \
+		ittt ne; \
+		ldrbne r3, [%1], #1; \
+		strbne r3, [%0], #1; \
+		subsne %2, #1; \
+		bne 2b"
+	: "+r" (dst), "+r" (src), "+r" (l)
+	:
+	: "r3", "memory", "cc");
+
+}
+
+
+int hal_memcmp(const void *ptr1, const void *ptr2, size_t num)
+{
+	int res = 0;
+
+	__asm__ volatile
+	(" \
+	1: \
+		cmp %3, #0; \
+		beq 3f; \
+		sub %3, #1; \
+		ldrb r3, [%1], #1; \
+		ldrb r4, [%2], #1; \
+		cmp r3, r4; \
+		beq 1b; \
+		blo 2f; \
+		mov %0, #1; \
+		b 3f; \
+	2: \
+		mov %0, #-1; \
+	3: "
+	: "+r" (res), "+r" (ptr1), "+r" (ptr2), "+r" (num)
+	:
+	: "r3", "r4", "memory", "cc");
+
+	return res;
+}
+
+
+void hal_memset(void *dst, int v, unsigned int l)
+{
+	unsigned int v1 = v & 0xffu;
+	unsigned int tmp;
+
+	tmp = (v1 << 8) | v1;
+	tmp |= (tmp << 16u);
+
+	__asm__ volatile
+	(" \
+		lsls r3, %0, #30; \
+		bne 2f; \
+	1: \
+		cmp %2, #4; \
+		itt hs; \
+		strhs %1, [%0], #4; \
+		subshs %2, #4; \
+		bhs 1b; \
+	2: \
+		cmp %2, #0; \
+		itt ne; \
+		strbne %1, [%0], #1; \
+		subsne %2, #1; \
+		bne 2b"
+	: "+r"(dst), "+r" (tmp), "+r" (l)
+	:
+	: "r3", "memory", "cc");
+}
+
+
+unsigned int hal_strlen(const char *s)
+{
+	unsigned int k = 0;
+
+	__asm__ volatile
+	(" \
+	1: \
+		ldrb r1, [%1, %0]; \
+		cbz r1, 2f; \
+		add %0, #1; \
+		b 1b; \
+	2:"
+	: "+r" (k), "+r" (s)
+	:
+	: "r1", "memory", "cc");
+
+	return k;
```

Comment:
This function can be improved, i.e., one instruction can be reduced in the loop.
If you are interested I can present it.

Reply:
What do you mean?